myia/ir.py

```diff
@@ -49,22 +53,23 @@ class Node:
     def __init__(self) -> None:
         """Construct a node."""
         self.inputs: List[Node] = []
-        self.value: LiteralType = None
+        self.uses: Optional[List[Node]] = None
```

Comment:
At a minimum, type should be `Optional[Set[Tuple[int, Node]]]`, because we need to know which of the predecessor's inputs points to self (unless we want to loop through the whole list).

I would consider being a bit more general, though: it could be useful to put `(OUT, Function)` in the set of uses of a function's graph's output so that we can have the property that a node is dead when its use set is empty.

Alternatively, we might want an `Edge` type. I'll post about that later.

Reply:
What's wrong with looping through the entire list? Seems like a bit of premature optimization to me to keep a set of tuples around. Looping through a tiny list is literally nanoseconds, and only tests that involve knowing whether you are the *n*-th input to a node are faster this way. There are lots of other tests which are theoretically slower this way because you are walking through a hash table instead of a list.

`(OUT, Function)` seems hacky to me. You can just prune unused nodes by walking from a graph's outputs to inputs and throw out all unvisisted nodes in one go. The `(OUT, Function)` trick would have to be applied iteratively instead.

Reply:
I mean, I think it's a bigger pain to loop through that list than to have tuples. The use case is node replacement in general: you loop through A's uses, then you replace each of them with a pointer to B. If you store node and index in a tuple, it's one simple loop, if you only store the node, it's two nested loops with a check. This isn't a fringe case, this is basically all cases. When would you loop over predecessors without needing the indices?

`(OUT, Function)` makes sense semantically, insofar that it makes sense to draw an edge of sorts from a Function to its output. You can prune unused nodes in one pass, but doing it incrementally could turn out to be useful as well. It's also a way to locally store the information needed to replace an output instead of having to loop over all graphs.

Reply:
That pattern could be implemented in a single method `replace` which implements that nested loop. The set of tuples approach introduces redundant information as well e.g. if I change the order of inputs I must be sure to update the inputs' `uses` attributes to reflect this.

Example of where you don't need the indices: I want to find all the addition nodes where I am an input, and the other input is a scalar, and I don't care about whether I am the first or second input, since addition is commutative.

I do agree that it should be a set rather than a list though, since the ordering is irrelevant. I'll change that.

If there is a case where we need to identify a node as being an output easily, why not just add a boolean `is_output` attribute?  Since`uses` is supposed to be the reverse of `inputs`, why add some special case where one of the elements is not even a `Node` instance, but something else entirely.

Reply:
> The set of tuples approach introduces redundant information as well e.g. if I change the order of inputs I must be sure to update the inputs' `uses` attributes to reflect this.

The interface you should be using to change inputs should do this for you. Tuples or not, handling the bookkeeping for uses yourself instead of going through a safe API is a recipe for disaster. Sure, you know it's safe when you swap two inputs, but you're still playing with fire.

> Example of where you don't need the indices: I want to find all the addition nodes where I am an input, and the other input is a scalar, and I don't care about whether I am the first or second input, since addition is commutative.

I mean, the index does help a little in this case, because it gives you the index of the other input directly:

```python
for pred in node.uses:
    if isadd(pred.inputs[0]) and any(constant(x) for x in pred.inputs[1:] if x != node):
        ...
```

vs

```python
for i, pred in node.uses:
    if isadd(pred.inputs[0]) and constant(pred.inputs[3-i]):
        ...
```

> If there is a case where we need to identify a node as being an output easily, why not just add a boolean `is_output` attribute?

You don't just want to identify it as an output, you want to identify the graph it is an output of: if you replace A by B, and A is the output of graph G, you need to change G's output pointer to point to B instead. I'll also note that a node could be, in theory, the output of more than one graph, but we probably want to merge them in that situation.

> Since `uses` is supposed to be the reverse of `inputs`

I mean that's the point, I don't necessarily agree that this is all what it is supposed to be. I view it more as a repository of what points to this node and should be adjusted if this node is replaced.